utils/src/main/java/software/amazon/awssdk/utils/async/AbstractFlatteningSubscriber.java

```diff
@@ -0,0 +1,274 @@
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.utils.async;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.utils.Logger;
+
+@SdkInternalApi
+public abstract class AbstractFlatteningSubscriber<T, U> extends DelegatingSubscriber<T, U> {
+    private static final Logger log = Logger.loggerFor(AbstractFlatteningSubscriber.class);
+
+    /**
+     * The amount of unfulfilled demand open against the upstream subscriber.
+     */
+    protected final AtomicLong upstreamDemand = new AtomicLong(0);
+
+    /**
+     * The amount of unfulfilled demand the downstream subscriber has opened against us.
+     */
+    protected final AtomicLong downstreamDemand = new AtomicLong(0);
+
+    /**
+     * A flag that is used to ensure that only one thread is handling updates to the state of this subscriber at a time. This
+     * allows us to ensure that the downstream onNext, onComplete and onError are only ever invoked serially.
+     */
+    protected final AtomicBoolean handlingStateUpdate = new AtomicBoolean(false);
+
+    /**
+     * Whether the upstream subscriber has called onError on us. If this is null, we haven't gotten an onError. If it's non-null
+     * this will be the exception that the upstream passed to our onError. After we get an onError, we'll call onError on the
+     * downstream subscriber as soon as possible.
+     */
+    protected final AtomicReference<Throwable> onErrorFromUpstream = new AtomicReference<>(null);
+
+    /**
+     * Whether we have called onComplete or onNext on the downstream subscriber.
+     */
+    protected volatile boolean terminalCallMadeDownstream = false;
+
+    /**
+     * Whether the upstream subscriber has called onComplete on us. After this happens, we'll drain any outstanding items in the
+     * allItems queue and then call onComplete on the downstream subscriber.
+     */
+    protected volatile boolean onCompleteCalledByUpstream = false;
+
+    /**
+     * The subscription to the upstream subscriber.
+     */
+    protected Subscription upstreamSubscription;
+
+    protected AbstractFlatteningSubscriber(Subscriber<? super U> subscriber) {
+        super(subscriber);
+    }
+
+    /**
+     * This method is called inside the onNext signal. Implementation should do what is required to store the data
+     * before fulfilling the demand from the downstream subscriber.
+     *
+     * @param item the value with which onNext was called.
+     */
+    protected abstract void doWithItem(T item);
```

Comment:
I can't think of a better naming for this one, can we make this package private so that we can change it in the future?